bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/relay/RelayHandler.java

```diff
@@ -0,0 +1,296 @@
+    }
+
+    private Channel createChannel(String channelId) {
+        ChannelUID channelUID = new ChannelUID(getThing().getUID(), channelId);
+        ChannelTypeUID channelTypeUID = getChannelTypeUID(channelId);
+        return ChannelBuilder.create(channelUID).withType(channelTypeUID).build();
+    }
+
+    private ChannelTypeUID getChannelTypeUID(String channelId) {
+        switch (channelId) {
+            case CHANNEL_IMPULSE_SWITCH, CHANNEL_IMPULSE_LENGTH, CHANNEL_INSTANT_OF_LAST_IMPULSE:
+                return new ChannelTypeUID(BINDING_ID, channelId);
+            case CHANNEL_POWER_SWITCH:
+                return new ChannelTypeUID("system:power");
+            default:
+                throw new UnsupportedOperationException(
+                        "Cannot determine channel type UID to create channel " + channelId + " dynamically.");
+
+        }
+    }
+
+    private boolean isRelayInImpulseSwitchMode(Device deviceInfo) {
+        List<String> serviceIds = deviceInfo.deviceServiceIds;
+        return serviceIds != null && serviceIds.contains(ImpulseSwitchService.IMPULSE_SWITCH_SERVICE_NAME);
+    }
+
+    @Override
+    protected void initializeServices() throws BoschSHCException {
+        if (!isInImpulseSwitchMode) {
+            // initialize PowerSwitch service only if the relay is not configured as impulse switch
+            super.initializeServices();
+        } else {
+            // initialize impulse switch service only if the relay is configured as impulse switch
+            registerService(impulseSwitchService, this::updateChannels,
+                    List.of(CHANNEL_IMPULSE_SWITCH, CHANNEL_IMPULSE_LENGTH, CHANNEL_INSTANT_OF_LAST_IMPULSE), true);
+        }
+
+        createService(CommunicationQualityService::new, this::updateChannels, List.of(CHANNEL_SIGNAL_STRENGTH), true);
+        registerService(childProtectionService, this::updateChannels, List.of(CHANNEL_CHILD_PROTECTION), true);
+    }
+
+    private void updateChannels(CommunicationQualityServiceState communicationQualityServiceState) {
+        updateState(CHANNEL_SIGNAL_STRENGTH, communicationQualityServiceState.quality.toSystemSignalStrength());
+    }
+
+    private void updateChannels(ChildProtectionServiceState childProtectionServiceState) {
+        updateState(CHANNEL_CHILD_PROTECTION, OnOffType.from(childProtectionServiceState.childLockActive));
+    }
+
+    private void updateChannels(ImpulseSwitchServiceState impulseSwitchServiceState) {
+        this.currentImpulseSwitchServiceState = impulseSwitchServiceState;
+
+        updateState(CHANNEL_IMPULSE_SWITCH, OnOffType.from(impulseSwitchServiceState.impulseState));
+        updateState(CHANNEL_IMPULSE_LENGTH, new DecimalType(impulseSwitchServiceState.impulseLength));
+
+        if (impulseSwitchServiceState.instantOfLastImpulse != null) {
+            updateState(CHANNEL_INSTANT_OF_LAST_IMPULSE,
+                    new DateTimeType(impulseSwitchServiceState.instantOfLastImpulse));
+        }
```

Comment:
Is it possible that this changes from != null to == null? In that case we might want to clear the state?
```suggestion
        updateState(CHANNEL_INSTANT_OF_LAST_IMPULSE, impulseSwitchServiceState.instantOfLastImpulse != null
                ? new DateTimeType(impulseSwitchServiceState.instantOfLastImpulse))
                : UnDefType.NULL;
```

Reply:
I don't know if this can happen because I don't have the hardware to test. I guess once it's non-null it logically should not go back to `null`. But we can still add the code just ist case :+1: